### ISSUE TITLE
Add to configuration files to define requirements for ESSGlobal dictionaries

### ISSUE THREAD

**Issue by ColmMassey** [SolidarityEconomyAssociation/sea-map#104]
Define the structure of new config file  (or extension of existing one) which specifies which vocabularies are required by this version of sea-map and, what sparql query is needed to generate that dictionary. 
We also need to define which of these vocabularies will be made available as filter options to the advanced search panel.
 
The main extension will look something like...

```
{  
  "structure-ICA": "
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX osspatialrelations: <http://data.ordnancesurvey.co.uk/ontology/spatialrelations/>
    SELECT  ?termURI ?termLabel
    WHERE {
	  ?uri rdf:type essglobal:SSEInitiative .
	  ?uri gr:name ?name .
    }
  ",
  "activities-ICA": "
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX osspatialrelations: <http://data.ordnancesurvey.co.uk/ontology/spatialrelations/>
    SELECT  ?termURI ?termLabel
    WHERE {
	  ?uri rdf:type essglobal:SSEInitiative .
	  ?uri gr:name ?name .
    }
  ",
  ...  
  "super-region-ICA": "
    PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
    PREFIX osspatialrelations: <http://data.ordnancesurvey.co.uk/ontology/spatialrelations/>
    SELECT  ?termURI ?termLabel
    WHERE {
	  ?uri rdf:type essglobal:SSEInitiative .
	  ?uri gr:name ?name .
    }
  "
}

```
We already have the following list in config.json

```
"filterableFields": [
  {
    "field": "primaryActivity",
    "label": "Activities"
  }
]
```

which specifies the primary field in the directory. We could extend this to a list of vocabularies, where the first in line is the one used in the directory.

```
"filterableFields": [
    "country-ICA",
    "structure-ICA",
    "activities-ICA",
    "topology-ICA",
    "region-ICA",
    "super-region-ICA"
]
```

**Comment by ColmMassey** [comment]
Note, we dropped the label as this needs to be provided within the dictionaries.

Does this look about right @King-Mob & @wu-lee ?

**Comment by wu-lee** [comment]
*The TL;DR here is: I'm guessing we need to supply more or less what's already being used in sea-map (see below), and so here's a proposal how to get there, using minimal config and a SPARQL query template. We could supply literal queries as above, not totally against this, but I am wondering if that's going to just make life more complicated - more cut & paste of identical queries - with little gain now that we're not planning on using the `get_datasets.php` mechanism as-is, which uses literal queries as config?*

My sketchy outline for the new `config.json` parameters doesn't include the queries, assuming that they can all be constructed in the PHP script `get_vocabs.php` from the config:

```
{
  ...
  "languages": ["EN", "ES", "FR"],
  "vocabularies": [
    "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/activities",
    "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure",
    "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/base-membership-type"
  ]
  ...
}
```

The template query for a vocab with URI `$uri` and language `$lang` would be of the form (approx guess here minus real prefixes):

```
SELECT  ?term ?label
WHERE {
    ?term whatever:belongsToVocab <$uri> .
    ?term whatever:hasLabel ?label .
    ?label whatever:isLang <$lang> .
}
```

Which returns a (JSON) datastructure of the form:
```
[
  ["ALL",  "All Activities"],
  ["AM10", "Arts, Media Culture & Leisure"],
  ...
  ["BMT60", "Others"]
]
```

Which can be munged by the PHP script into a datastructure pretty much exactly like the one already hard-wired into `sse_initiatives.js` (although I confess I don't know how to sensibly incorporate the `qualifier` thing yet). The existing one is the same but has English labels like "Activities" instead of URIs like "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/activities". 

I was thinking we might add an extra level so we can include more than one language in this index, which would mean querying for all languages above, but the basic idea is the same.

```
{
  "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/activities": {
    "ALL": "All Activities",
    "AM10": "Arts, Media, Culture & Leisure",
    "AM20": "Campaigning, Activism & Advocacy",
    "AM30": "Community & Collective Spaces",
    "AM40": "Education",
    "AM50": "Energy",
    "AM60": "Food",
    "AM70": "Goods & Services",
    "AM80": "Health, Social Care & Wellbeing",
    "AM90": "Housing",
    "AM100": "Money & Finance",
    "AM110": "Nature, Conservation & Environment",
    "AM120": "Reduce, Reuse, Repair & Recycle",
    "AM130": "Agriculture",
    "AM140": "Industry",
    "AM150": "Utilities",
    "AM160": "Transport",
    "ICA10": "Agriculture",
    "ICA20": "Dairy",
    "ICA30": "Forestry",
    "ICA40": "Irrigation",
    "ICA50": "Fishing",
    "ICA60": "Artisans",
    "ICA70": "Construction",
    "ICA80": "Industry",
    "ICA90": "Manufacturing",
    "ICA100": "Mining",
    "ICA110": "Professional",
    "ICA120": "Service",
    "ICA130": "Tourism",
    "ICA140": "Financial Services",
    "ICA150": "Insurance",
    "ICA160": "Education",
    "ICA170": "Health",
    "ICA180": "Community",
    "ICA190": "Social",
    "ICA200": "Social Service",
    "ICA210": "Housing",
    "ICA220": "Transport",
    "ICA230": "Utilities",
    "ICA240": "Retail",
    "ICA250": "Production",
    "ICA260": "Wholesale and retail trade",
    "ICA270": "Education / health / social work",
    "ICA280": "Other services",
    "ICA290": "All",
    "q09": "Coop Promoter/Supporter"
  },
  "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure": {
    "OS10": "Community group (formal or informal)",
    "OS20": "Not-for-profit organisation",
    "OS30": "Social enterprise",
    "OS40": "Charity",
    "OS50": "Company (Other)",
    "OS60": "Workers co-operative",
    "OS70": "Housing co-operative",
    "OS80": "Consumer co-operative",
    "OS90": "Producer co-operative",
    "OS100": "Multi-stakeholder co-operative",
    "OS110": "Secondary co-operative",
    "OS115": "Co-operative",
    "OS120": "Community Interest Company (CIC)",
    "OS130": "Community Benefit Society / Industrial and Provident Society (IPS)",
    "OS140": "Employee trust",
    "OS150": "Self-employed",
    "OS160": "Unincorporated",
    "OS170": "Mutual Organisation",
    "OS180": "National apex",
    "OS190": "National sectoral federation or union",
    "OS200": "Regional, state or provincial level federation or union",
    "OS210": "Cooperative group",
    "OS220": "Government agency/body",
    "OS230": "Supranational",
    "OS240": "Cooperative of cooperatives / mutuals"
  },
  "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/base-membership-type": {
    "BMT10": "Consumer/Users",
    "BMT20": "Producers",
    "BMT30": "Workers",
    "BMT40": "Multi-stakeholders",
    "BMT50": "Residents",
    "BMT60": "Others"
  }
};
```

**Comment by ColmMassey** [comment]
We need to test whether we can do it all with one sparql query template. So I'll start writing the with the qualifier exception.

**Comment by ColmMassey** [comment]
We need to add the vocabularies to virtuoso. At the moment I think only https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure has been added.

**Comment by ColmMassey** [comment]
I am struggling with the different spellings of Organisational Structure

It appears as _OrganisationalStructure_ _organisationalStructure_ and _organisational-structure_

**Comment by wu-lee** [comment]
> We need to add the vocabularies to virtuoso. At the moment I think only https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure has been added.

Yes - I'll see if I can do that. But I was thinking we should prioritise getting a stub returning some canned data first?

**Comment by ColmMassey** [comment]
PREFIX essgvocab: <https://w3id.solidarityeconomy.coop/essglobal/V2a/vocab>
PREFIX essgstandard: <https://w3id.solidarityeconomy.coop/essglobal/V2a/standard>
PREFIX lod: <https://lod.coop/>

lod:sea-lod/newbridge/118 essgvocab:organisationalStructure essgstandard:organisational-structure/OS40

<nodeID://b3435684> <http://www.w3.org/2002/07/owl#hasValue> essgstandard:organisational-structure

essgvocab:OrganisationalStructure <http://www.w3.org/2000/01/rdf-schema#label> 'Organisational Structure VES'
essgvocab:organisationalStructure  <http://www.w3.org/2000/01/rdf-schema#label> 'Organisational Structure'

essgvocab:organisationalStructure <http://www.w3.org/2000/01/rdf-schema#range> essgvocab:OrganisationalStructure

**Comment by ColmMassey** [comment]
Previous comment was just a list of all types of occurences of the many spellings of "organisational structure" currently found in graphs.

**Comment by ColmMassey** [comment]
> Yes - I'll see if I can do that. But I was thinking we should prioritise getting a stub returning some canned data first?

For sure.

**Comment by wu-lee** [comment]
> I am struggling with the different spellings of Organisational Structure
> It appears as _OrganisationalStructure_ _organisationalStructure_ and _organisational-structure_

I noticed that too - I think the convention in linked data is:
- Class names are CamelCase (start with a capital) and are subjects or objects: so `essglobal:OrganisationalStructure` is a subclass of `skos:Concept`
- Properties are camelCase (start with lower-case) and are predicates: so `<someURI>` might have an `essglobal:organisationalStructure` property `<https://blah/OS10>`
- URIs need (ought) to be lower-case, because strictly are case-insensitive, so "snake-case" is used

I tried to write a query which lists terms in a skos vocabulary on DBPedia, but I noticed that I couldn't replicate it to our data because the relationships are different. I think:
- In DBpedia, vocabs have `rdf:type`  `skos:Concept`
- In ESSGlobal, vocabs are `rdfs:subClassOf` `skos:Concept`

I'm not sure they're compatible, as you'd think that 'rdf:type` is an edge case of `rdfs:subClassOf` but I couldn't get data out for that on ESSGlobal data. So:
- Only gets results in ESSGlobal: `select * where {?x rdfs:subClassOf skos:Concept  . }`
- Only gets results on DBpedia: `select * where {?x rdf:type skos:Concept  . }`

**Comment by ColmMassey** [comment]
> * Class names are CamelCase (start with a capital) and are subjects or objects: so `essglobal:OrganisationalStructure` is a subclass of `skos:Concept`
> 
> * Properties are camelCase (start with lower-case) and are predicates: so `<someURI>` might have an `essglobal:organisationalStructure` property `<https://blah/OS10>`
> 
> * URIs need (ought) to be lower-case, because strictly are case-insensitive, so "snake-case" is used

Yes that makes sense

**Comment by ColmMassey** [comment]
> `qualifier` thing yet

The sparql query needs to return a dictionary match like

"q09": "Coop Promoter/Supporter"

included in the vocab returned for the structures.
so something like 

  "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structureICA": {
    "OS115": "Co-operative",
    "OS170": "Mutual Organisation",
    "OS180": "National apex",
    "OS190": "National sectoral federation or union",
    "OS200": "Regional, state or provincial level federation or union",
    "OS210": "Cooperative group",
    "OS240": "Cooperative of cooperatives / mutuals"
    "q09": "Coop Promoter/Supporter"
  },

but as the qualifiers are in different vocabs, we may need to include a bit more of the uri, so more like

  "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structureICA": {
     "title": "Structure Type",
    "terms": {
      "organisational-structure/OS115": "Co-operative",
      "organisational-structure/OS170": "Mutual Organisation",
      "organisational-structure/OS180": "National apex",
      "organisational-structure/OS190": "National sectoral federation or union",
      "organisational-structure/OS200": "Regional, state or provincial level federation or union",
      "organisational-structure/OS210": "Cooperative group",
      "organisational-structure/OS240": "Cooperative of cooperatives / mutuals",
      "qualifier/q09": "Coop Promoter/Supporter",
    }
  }

**Comment by wu-lee** [comment]
It does seem odd to smoosh all these vocabs into one in what seems like an arbitrary way, strictly they are different namespaces and we're plonking them all into one, just lucky that they have different prefixes like `OS` or `Q`

And I'm wondering why is it lower-case `q09` yet upper case `OS240`?

**Comment by ColmMassey** [comment]
> It does seem odd to smoosh all these vocabs into one in what seems like an arbitrary way

I think it is more including terms from different 'standard' vocabularies into a new client defined vocabulary.

**Comment by ColmMassey** [comment]
> And I'm wondering why is it lower-case `q09` yet upper case `OS240`?

Yup, one of several inconsistencies.

**Comment by ColmMassey** [comment]
`> "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structureICA": {`

In the spirit of consistency, let'sreplace this with 

`"https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structure-ica": {`

**Comment by wu-lee** [comment]
> > It does seem odd to smoosh all these vocabs into one in what seems like an arbitrary way
> 
> I think it is more including terms from different 'standard' vocabularies into a new client defined vocabulary.

In which case, don't we need a official URI for the client-defined vocab? Presumably in this case it would be `https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structure-ica`, but as it stands I don't think that vocab currently includes any term for Q09)? 

Also, since any term could be included, no common prefix guaranteed, it essentially forces us to use the whole URI as the key value? So:

```
"https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structureICA": {
    "title": "Structure Type",
   "terms": {
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS115": "Co-operative",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS170": "Mutual Organisation",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS180": "National apex",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS190": "National sectoral federation or union",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS200": "Regional, state or provincial level federation or union",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS210": "Cooperative group",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS240": "Cooperative of cooperatives / mutuals",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/qualifier#q09": "Coop Promoter/Supporter"
   }
 }
```

Unless perhaps we allow SPARQL style prefix notation:
```
"https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/structureICA": {
    "title": "Structure Type",
   "prefixes": {
      "os": "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure",
      "q": "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/qualifier"
   },
   "terms": {
     "os:OS115": "Co-operative",
     "os:OS170": "Mutual Organisation",
     "os:OS180": "National apex",
     "https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/organisational-structure#OS190": "National sectoral federation or union",
     "os:OS200": "Regional, state or provincial level federation or union",
     "os:OS210": "Cooperative group",
     "os:OS240": "Cooperative of cooperatives / mutuals",
     "q:q09": "Coop Promoter/Supporter"
   }
 }
```

**Comment by ColmMassey** [comment]
This file are what I think should work for structure, activitues and typology.

[Dictionary.txt](https://github.com/SolidarityEconomyAssociation/sea-map/files/5967751/Dictionary.txt)

**Comment by wu-lee** [comment]
I've added that Dictionary.txt to a gist [here][1] so we can update it. I've also fixed some JSON syntax errors I noticed, and added scripts mapping it to what I think the originating SPARQL query result will look like ([terms.json][]), and back again to [vocabs2.json][], which has some proposed amendments, and has been manually edited to add titles back (see point below) and more sensible prefix abbreviations. 

Taking a close look at this and our proposal so far, some proposals / observations:
- To avoid duplication perhaps we should we move the multiple `prefixes` sections up to the top level and merge them into one? Although this means to avoid it being at the same level as vocab identifiers, we might need to have a `{ vocabs: { ... }, prefixes: { ... } }` structure.
- We may need to include prefixes for terms *and* vocabs? e.g. The vocab for StructureICA includes terms from OrganisationalStructure, plus a term from Qualifier.
- Some vocabs have an "All" term, some don't.  This might need to be made consistent. When there is no "All" term, sea-map's hardwired list has added a synthetic one, but which seems to be cross-vocabulary. So in general: "All" can mean more than one thing, and we should be careful.
- Vocab titles either need a fifth field in the results (with much duplication therefore), or a separate query for titles (see [titles.json][]).

The [vocabs2.json][] file is an example of these proposals. It has a `prefixes` section which applies to both vocab and term URIs, and I have adjusted the nesting. See what you think.

[1]: https://gist.github.com/wu-lee/af0618d567e5235b21dbd5fdd2674075
[vocabs.json]: https://gist.github.com/wu-lee/af0618d567e5235b21dbd5fdd2674075#file-vocabs-json
[vocabs2.json]: https://gist.github.com/wu-lee/af0618d567e5235b21dbd5fdd2674075#file-vocabs2-json
[terms.json]: https://gist.github.com/wu-lee/af0618d567e5235b21dbd5fdd2674075#file-terms-json
[titles.json]: https://gist.github.com/wu-lee/af0618d567e5235b21dbd5fdd2674075#file-titles-json

**Comment by ColmMassey** [comment]
As is decided not to investigate gists yet, here is the country dictionary. The sausage factory will need to be changed to use these terms. I have called it essglobal:ISO-3166 as it is very close to an exact subset of ISO-3166-1 and this is what we should move to support soon. (I can find and xml schema which purports to be rdf, but I couldn't load it.)

We will need to create ghost uris (uris with no urls for them), for the moment, for countries, as I want to research further the best uri to use for this.

something like prefix coun: <https://w3id.solidarityeconomy.coop/essglobal/V2a/standard/country/>

**Comment by ColmMassey** [comment]
[Regions-super-regions.txt](https://github.com/SolidarityEconomyAssociation/sea-map/files/5972939/Regions-super-regions.txt)

**Comment by wu-lee** [comment]
ISO seem to want money to get their list of countries, although copies exist elsewhere, as you say, I can't find any linked-data examples which work save perhaps this: https://www.omg.org/spec/LCC/Countries/ISO3166-1-CountryCodes/

That doesn't render in my browser, however there does seem to be RDF data there.

**Comment by wu-lee** [comment]
In your vocab, your country identifiers are not two letters. Doesn't ISO3166-1 define a two-letter code?  So I infer this is the ICA's list?

So perhaps we should shorten them to match the ISO identifiers?

**Comment by ColmMassey** [comment]
> So perhaps we should shorten them to match the ISO identifiers?

We could. It would be another step in the right direction. Shorter and language free.

**Comment by ColmMassey** [comment]
> Doesn't ISO3166-1 define a two-letter code?

And 3 letter, which make guessing the country from the code easier for humans.

**Comment by wu-lee** [comment]
I've integrated/tidied the country vocabs into `vocabs.json` in the gist, and re-generated `terms.json`, `titles.json` and `vocabs2.json`
